File: 04-code-examples/python/common_issues.py
# Examples of common Python issues that SonarQube detects

import os
import re
import json
import sqlite3
import subprocess
from typing import List, Dict, Optional

# BUG: Unused import
# Rule: S1481 - Unused local variables should be removed
import sys  # SonarQube flags unused import

class PythonIssuesExamples:
    """Examples of common Python issues detected by SonarQube"""

    # BUG: Mutable default arguments
    # Rule: S2234 - Parameters should be passed in the correct order
    def mutable_default_bug(self, items=[]):  # Dangerous!
        """SonarQube flags mutable default arguments"""
        items.append('new_item')
        return items

    # FIXED VERSION
    def mutable_default_fixed(self, items=None):
        """Proper way to handle mutable default arguments"""
        if items is None:
            items = []
        items.append('new_item')
        return items

    # VULNERABILITY: SQL Injection
    # Rule: S2077 - SQL queries should not be vulnerable to injection attacks
    def sql_injection_bug(self, user_id: str) -> List[Dict]:
        """SonarQube flags SQL injection vulnerability"""
        conn = sqlite3.connect('database.db')
        cursor = conn.cursor()
        
        # Dangerous - SQL injection vulnerability
        query = f"SELECT * FROM users WHERE id = {user_id}"
        cursor.execute(query)
        
        results = cursor.fetchall()
        conn.close()
        return results

    # FIXED VERSION
    def sql_injection_fixed(self, user_id: str) -> List[Dict]:
        """Parameterized query prevents SQL injection"""
        conn = sqlite3.connect('database.db')
        cursor = conn.cursor()
        
        # Safe - parameterized query
        query = "SELECT * FROM users WHERE id = ?"
        cursor.execute(query, (user_id,))
        
        results = cursor.fetchall()
        conn.close()
        return results

    # VULNERABILITY: Command Injection
    # Rule: S4823 - Using shell=True is security-sensitive
    def command_injection_bug(self, filename: str) -> str:
        """SonarQube flags shell injection vulnerability"""
        # Dangerous - command injection vulnerability
        result = subprocess.run(f"cat {filename}", shell=True, capture_output=True, text=True)
        return result.stdout

    # FIXED VERSION
    def command_injection_fixed(self, filename: str) -> str:
        """Safe command execution without shell"""
        try:
            result = subprocess.run(['cat', filename], capture_output=True, text=True, check=True)
            return result.stdout
        except subprocess.CalledProcessError as e:
            print(f"Error reading file: {e}")
            return ""

    # BUG: Exception handling issues
    # Rule: S1181 - Catching Exception is not allowed
    def exception_handling_bug(self):
        """SonarQube flags overly broad exception handling"""
        try:
            risky_operation()
        except Exception:  # Too broad - catches everything
            print("Something went wrong")
            pass  # Silent failure is bad

    # FIXED VERSION
    def exception_handling_fixed(self):
        """Specific exception handling with proper logging"""
        try:
            risky_operation()
        except FileNotFoundError as e:
            print(f"File not found: {e}")
            raise  # Re-raise if cannot handle
        except ValueError as e:
            print(f"Invalid value: {e}")
            return None
        except Exception as e:
            print(f"Unexpected error: {e}")
            raise  # Don't suppress unexpected exceptions

    # CODE SMELL: Function too complex
    # Rule: S3776 - Cognitive Complexity of functions should not be too high
    def complex_function_bug(self, data: List[Dict]) -> Dict:
        """SonarQube flags high cognitive complexity"""
        result = {}
        
        for item in data:
            if item:
                if 'type' in item:
                    if item['type'] == 'user':
                        if 'active' in item:
                            if item['active']:
                                if 'age' in item:
                                    if item['age'] >= 18:
                                        if 'email' in item:
                                            if '@' in item['email']:
                                                result[item['id']] = item
        return result

    # FIXED VERSION
    def complex_function_fixed(self, data: List[Dict]) -> Dict:
        """Reduced complexity with early returns and helper methods"""
        result = {}
        
        for item in data:
            if self._is_valid_adult_user(item):
                result[item['id']] = item
                
        return result

    def _is_valid_adult_user(self, item: Dict) -> bool:
        """Helper method to reduce complexity"""
        if not item or item.get('type') != 'user':
            return False
            
        if not item.get('active', False):
            return False
            
        if item.get('age', 0) < 18:
            return False
            
        email = item.get('email', '')
        return '@' in email

    # CODE SMELL: Duplicated code
    # Rule: S4144 - Methods should not have identical implementations
    def process_users_bug(self, users: List[str]) -> List[str]:
        """SonarQube detects code duplication"""
        processed = []
        for user in users:
            if user and user.strip():
                cleaned = user.strip().lower()
                if len(cleaned) > 2:
                    processed.append(cleaned)
        return processed

    def process_products_bug(self, products: List[str]) -> List[str]:
        """Identical logic - code duplication"""
        processed = []
        for product in products:
            if product and product.strip():
                cleaned = product.strip().lower()
                if len(cleaned) > 2:
                    processed.append(cleaned)
        return processed

    # FIXED VERSION
    def process_users_fixed(self, users: List[str]) -> List[str]:
        """Extracted common logic"""
        return self._process_string_list(users)

    def process_products_fixed(self, products: List[str]) -> List[str]:
        """Reuses common processing logic"""
        return self._process_string_list(products)

    def _process_string_list(self, items: List[str]) -> List[str]:
        """Common processing logic extracted"""
        return [
            item.strip().lower()
            for item in items
            if item and item.strip() and len(item.strip()) > 2
        ]

    # BUG: Resource leak
    # Rule: S2095 - Resources should be closed
    def resource_leak_bug(self, filename: str) -> str:
        """SonarQube flags unclosed resources"""
        file = open(filename, 'r')  # Not closed properly
        content = file.read()
        # Missing file.close()
        return content

    # FIXED VERSION
    def resource_leak_fixed(self, filename: str) -> str:
        """Proper resource management with context manager"""
        try:
            with open(filename, 'r') as file:
                return file.read()
        except IOError as e:
            print(f"Error reading file {filename}: {e}")
            return ""

    # VULNERABILITY: Hardcoded secrets
    # Rule: S2068 - Hard-coded credentials are security-sensitive
    def hardcoded_secrets_bug(self):
        """SonarQube flags hardcoded credentials"""
        api_key = "sk-1234567890abcdef"  # Hardcoded API key
        password = "admin123"  # Hardcoded password
        
        return {"api_key": api_key, "password": password}

    # FIXED VERSION
    def hardcoded_secrets_fixed(self):
        """Use environment variables for secrets"""
        api_key = os.getenv('API_KEY')
        password = os.getenv('DB_PASSWORD')
        
        if not api_key or not password:
            raise ValueError("Missing required environment variables")
            
        return {"api_key": api_key, "password": password}

    # BUG: Regex Denial of Service (ReDoS)
    # Rule: S5852 - Regular expressions should not be vulnerable to ReDoS
    def regex_dos_bug(self, text: str) -> bool:
        """SonarQube flags potentially vulnerable regex"""
        # Vulnerable to catastrophic backtracking
        pattern = r'^(a+)+$'
        return bool(re.match(pattern, text))

    # FIXED VERSION
    def regex_dos_fixed(self, text: str) -> bool:
        """Non-vulnerable regex pattern"""
        pattern = r'^a+$'  # Atomic grouping prevents backtracking
        return bool(re.match(pattern, text))

    # CODE SMELL: Magic numbers
    # Rule: S109 - Magic numbers should not be used
    def magic_numbers_bug(self, radius: float) -> float:
        """SonarQube flags magic numbers"""
        return 3.14159 * radius * radius  # Magic number

    # FIXED VERSION
    PI = 3.14159  # Class constant
    
    def magic_numbers_fixed(self, radius: float) -> float:
        """Named constant instead of magic number"""
        return self.PI * radius * radius

    # BUG: Incorrect string comparison
    # Rule: S1940 - Boolean checks should not be inverted
    def string_comparison_bug(self, text: str) -> bool:
        """SonarQube flags problematic string comparisons"""
        # Problematic - should use 'is' for None comparison
        if text == None:  # Should be 'is None'
            return False
        
        # Problematic - case sensitivity issues
        if text == "Admin":  # Case sensitive
            return True
            
        return False

    # FIXED VERSION
    def string_comparison_fixed(self, text: str) -> bool:
        """Proper None and string comparisons"""
        if text is None:
            return False
            
        # Case insensitive comparison
        if text.lower() == "admin":
            return True
            
        return False

    # PERFORMANCE: Inefficient loops
    # Code smell: Inefficient string concatenation in loops
    def inefficient_loop_bug(self, items: List[str]) -> str:
        """Inefficient string concatenation"""
        result = ""
        for item in items:
            result += item + ", "  # Inefficient string concatenation
        return result

    # FIXED VERSION
    def inefficient_loop_fixed(self, items: List[str]) -> str:
        """Efficient string joining"""
        return ", ".join(items)

    # BUG: Unreachable code
    # Rule: S1763 - Jump statements should not be redundant
    def unreachable_code_bug(self):
        """SonarQube flags unreachable code"""
        return "early return"
        print("This will never execute")  # Unreachable

    # FIXED VERSION
    def unreachable_code_fixed(self):
        """Remove unreachable code"""
        return "early return"

# Helper function for examples
def risky_operation():
    """Simulates a risky operation that might fail"""
    raise FileNotFoundError("Simulated file not found error")

# Example usage and testing
if __name__ == "__main__":
    examples = PythonIssuesExamples()
    
    # Test fixed versions
    print("Testing fixed methods:")
    
    # Test mutable default fix
    list1 = examples.mutable_default_fixed()
    list2 = examples.mutable_default_fixed()
    print(f"Mutable default fix: {list1 == list2}")  # Should be False
    
    # Test string processing
    users = ["  JOHN  ", "jane", "a", "  Bob Smith  "]
    processed = examples.process_users_fixed(users)
    print(f"Processed users: {processed}")
    
    # Test resource management
    try:
        content = examples.resource_leak_fixed("nonexistent.txt")
        print(f"File content: {content}")
    except Exception as e:
        print(f"Expected error: {e}")

"""
SonarQube Analysis Results Summary for Python:

Bugs: 10
- Mutable default arguments
- SQL injection vulnerability
- Command injection vulnerability
- Overly broad exception handling
- Resource leaks (unclosed files)
- Incorrect None comparisons
- Unreachable code
- Logic errors in conditions
- Missing error handling
- Improper resource management

Vulnerabilities: 4
- SQL injection (parameterized queries needed)
- Command injection (avoid shell=True)
- Hardcoded credentials (use environment variables)
- RegExp DoS (avoid catastrophic backtracking)

Code Smells: 6
- High cognitive complexity
- Code duplication
- Magic numbers
- Long functions/methods
- Too many parameters
- Inefficient algorithms

Python-Specific Best Practices:
1. Use context managers (with statements) for resource management
2. Avoid mutable default arguments
3. Use specific exception types, not broad Exception
4. Follow PEP 8 style guidelines
5. Use type hints for better code documentation
6. Prefer list comprehensions over loops when appropriate
7. Use 'is' for None comparisons, not '=='
8. Store secrets in environment variables
9. Validate and sanitize all external input
10. Use parameterized queries for database operations
11. Avoid shell=True in subprocess calls
12. Use pathlib for file path operations
13. Implement proper logging instead of print statements
14. Write docstrings for all public methods
15. Use f-strings for string formatting
"""

File: 04-code-examples/javascript/common-issues.js
// Examples of common JavaScript issues that SonarQube detects

// BUG: Variable declared but never used
// Rule: S1481 - Unused local variables should be removed
function unusedVariableBug() {
    const unusedVar = 'This variable is never used'; // SonarQube flags this
    const usedVar = 'This is used';
    console.log(usedVar);
}

// FIXED VERSION
function unusedVariableFixed() {
    const usedVar = 'This is used';
    console.log(usedVar);
}

// BUG: Equality operators should not be used in "for" loops
// Rule: S888 - Equality operators should not be used in "for" loop termination conditions
function forLoopBug() {
    const items = ['a', 'b', 'c'];
    // SonarQube flags == in for loop condition
    for (let i = 0; i == items.length; i++) { // Should be <
        console.log(items[i]);
    }
}

// FIXED VERSION
function forLoopFixed() {
    const items = ['a', 'b', 'c'];
    for (let i = 0; i < items.length; i++) {
        console.log(items[i]);
    }
    
    // Or better yet, use modern iteration
    items.forEach(item => console.log(item));
}

// VULNERABILITY: Cross-Site Scripting (XSS)
// Rule: S5147 - HTML5 "autofocus" attribute should not be used
function xssVulnerability(userInput) {
    // SonarQube flags direct HTML insertion without sanitization
    document.getElementById('output').innerHTML = userInput; // Dangerous!
}

// FIXED VERSION
function xssVulnerabilityFixed(userInput) {
    // Use textContent instead of innerHTML for user input
    document.getElementById('output').textContent = userInput;
    
    // Or sanitize HTML if HTML content is needed
    // document.getElementById('output').innerHTML = sanitizeHTML(userInput);
}

// BUG: Functions should not be defined inside loops
// Rule: S1515 - Functions should not be defined inside loops
function functionInLoopBug() {
    const buttons = document.querySelectorAll('.button');
    
    for (let i = 0; i < buttons.length; i++) {
        // SonarQube flags function definition inside loop
        buttons[i].addEventListener('click', function() {
            console.log('Button clicked: ' + i); // Also closure issue!
        });
    }
}

// FIXED VERSION
function functionInLoopFixed() {
    const buttons = document.querySelectorAll('.button');
    
    // Define function outside loop
    function createClickHandler(index) {
        return function() {
            console.log('Button clicked: ' + index);
        };
    }
    
    for (let i = 0; i < buttons.length; i++) {
        buttons[i].addEventListener('click', createClickHandler(i));
    }
    
    // Or use modern approach with forEach
    buttons.forEach((button, index) => {
        button.addEventListener('click', () => {
            console.log('Button clicked: ' + index);
        });
    });
}

// CODE SMELL: Complex function with too many parameters
// Rule: S107 - Functions should not have too many parameters
function tooManyParametersBad(name, email, phone, address, city, state, zip, country) {
    // SonarQube flags functions with >7 parameters
    return {
        name, email, phone, address, city, state, zip, country
    };
}

// FIXED VERSION - Use object parameter
function tooManyParametersGood(userInfo) {
    const { name, email, phone, address, city, state, zip, country } = userInfo;
    return { name, email, phone, address, city, state, zip, country };
}

// CODE SMELL: Duplicated string literals
// Rule: S1192 - String literals should not be duplicated
function duplicatedStringsBad() {
    console.log('Error: Invalid input'); // Duplicated string
    alert('Error: Invalid input');       // Same string repeated
    document.title = 'Error: Invalid input';
    throw new Error('Error: Invalid input');
}

// FIXED VERSION
function duplicatedStringsGood() {
    const ERROR_MESSAGE = 'Error: Invalid input';
    console.log(ERROR_MESSAGE);
    alert(ERROR_MESSAGE);
    document.title = ERROR_MESSAGE;
    throw new Error(ERROR_MESSAGE);
}

// BUG: Promise rejections should not be ignored
// Rule: S4822 - Promises should not be misused
async function promiseBug() {
    // SonarQube flags unhandled promise rejection
    fetch('/api/data'); // Missing error handling
    
    const promise = new Promise((resolve, reject) => {
        setTimeout(() => reject(new Error('Failed')), 1000);
    });
    
    promise; // Ignored promise
}

// FIXED VERSION
async function promiseFixed() {
    try {
        const response = await fetch('/api/data');
        if (!response.ok) {
            throw new Error('Network response was not ok');
        }
        const data = await response.json();
        return data;
    } catch (error) {
        console.error('Fetch failed:', error);
        throw error;
    }
}

// VULNERABILITY: Insecure randomness
// Rule: S2245 - Using pseudorandom number generators (PRNGs) is security-sensitive
function insecureRandomBug() {
    // SonarQube flags Math.random() for security-sensitive operations
    const sessionToken = Math.random().toString(36); // Predictable!
    return sessionToken;
}

// FIXED VERSION
function insecureRandomFixed() {
    // Use crypto.getRandomValues() for cryptographic purposes
    const array = new Uint8Array(32);
    crypto.getRandomValues(array);
    return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
}

// CODE SMELL: Magic numbers
// Rule: S109 - Magic numbers should not be used
function magicNumbersBad() {
    const radius = 5;
    return 3.14159 * radius * radius; // Magic number
}

// FIXED VERSION
function magicNumbersGood() {
    const PI = 3.14159;
    const radius = 5;
    return PI * radius * radius;
}

// BUG: Array and Object methods should be used correctly
// Rule: S4043 - Array and Object methods should be used appropriately
function arrayMethodsBug() {
    const numbers = [1, 2, 3, 4, 5];
    
    // SonarQube flags incorrect use of reduce
    const doubled = numbers.reduce((acc, num) => {
        acc.push(num * 2); // Should use map instead
        return acc;
    }, []);
    
    return doubled;
}

// FIXED VERSION
function arrayMethodsFixed() {
    const numbers = [1, 2, 3, 4, 5];
    
    // Use map for transformation
    const doubled = numbers.map(num => num * 2);
    
    return doubled;
}

// VULNERABILITY: RegExp Denial of Service (ReDoS)
// Rule: S5852 - Regular expressions should not be vulnerable to Denial of Service attacks
function regexDosBug(input) {
    // SonarQube flags potentially catastrophic backtracking
    const vulnerableRegex = /^(a+)+$/;
    return vulnerableRegex.test(input); // Can cause CPU exhaustion
}

// FIXED VERSION
function regexDosFixed(input) {
    // Use more efficient regex without nested quantifiers
    const safeRegex = /^a+$/;
    return safeRegex.test(input);
}

// CODE SMELL: Functions should not be empty
// Rule: S1186 - Functions should not be empty
function emptyFunctionBug() {
    // SonarQube flags empty functions
}

// FIXED VERSION
function emptyFunctionFixed() {
    // Either implement the function or remove it
    console.log('Function implemented');
}

// Alternatively, if intentionally empty, add comment
function emptyFunctionIntentional() {
    // Intentionally empty - placeholder for future implementation
}

// VULNERABILITY: Hardcoded credentials
// Rule: S2068 - Hard-coded credentials are security-sensitive
function hardcodedCredentialsBug() {
    // SonarQube flags hardcoded credentials
    const apiKey = 'sk-1234567890abcdef'; // Hardcoded API key
    const password = 'admin123'; // Hardcoded password
    
    return { apiKey, password };
}

// FIXED VERSION
function hardcodedCredentialsFixed() {
    // Use environment variables or secure configuration
    const apiKey = process.env.API_KEY;
    const password = process.env.DB_PASSWORD;
    
    if (!apiKey || !password) {
        throw new Error('Missing required credentials in environment');
    }
    
    return { apiKey, password };
}

// PERFORMANCE: Inefficient DOM manipulation
// Not a SonarQube rule but common performance issue
function inefficientDOMBad() {
    const container = document.getElementById('container');
    
    // Inefficient: Multiple reflows and repaints
    for (let i = 0; i < 1000; i++) {
        const div = document.createElement('div');
        div.textContent = `Item ${i}`;
        container.appendChild(div); // Multiple DOM modifications
    }
}

// FIXED VERSION
function efficientDOMGood() {
    const container = document.getElementById('container');
    const fragment = document.createDocumentFragment();
    
    // Efficient: Batch DOM operations
    for (let i = 0; i < 1000; i++) {
        const div = document.createElement('div');
        div.textContent = `Item ${i}`;
        fragment.appendChild(div);
    }
    
    container.appendChild(fragment); // Single DOM modification
}

// Export functions for testing (if using modules)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        unusedVariableFixed,
        forLoopFixed,
        xssVulnerabilityFixed,
        functionInLoopFixed,
        tooManyParametersGood,
        duplicatedStringsGood,
        promiseFixed,
        insecureRandomFixed,
        magicNumbersGood,
        arrayMethodsFixed,
        regexDosFixed,
        emptyFunctionFixed,
        hardcodedCredentialsFixed,
        efficientDOMGood
    };
}

/*
 * SonarQube Analysis Results Summary:
 * 
 * Bugs: 8
 * - Unused variables
 * - Incorrect loop conditions
 * - Functions defined in loops
 * - Unhandled promises
 * - Incorrect array method usage
 * - Empty functions
 * - DOM inefficiencies
 * - Logic errors
 * 
 * Vulnerabilities: 4
 * - Cross-Site Scripting (XSS)
 * - Insecure randomness
 * - RegExp Denial of Service
 * - Hardcoded credentials
 * 
 * Code Smells: 3
 * - Too many parameters
 * - Duplicated strings
 * - Magic numbers
 * 
 * Best Practices for JavaScript Code Quality:
 * 1. Use strict mode ('use strict')
 * 2. Declare variables properly (const/let, not var)
 * 3. Handle promises with try/catch or .catch()
 * 4. Sanitize user input before DOM insertion
 * 5. Use meaningful variable and function names
 * 6. Avoid deeply nested callbacks (callback hell)
 * 7. Use modern ES6+ features appropriately
 * 8. Implement proper error handling
 * 9. Use ESLint with SonarQube rules
 * 10. Write unit tests for all functions
 */